lab4/lab4.rst

```diff
@@ -0,0 +1,230 @@
+------------------------------------
+
+You are given a file ``set_assoc.cc`` which has empty functions for each function you are required to implement.
+You will be modifying and turning in this file and ``set_assoc.hh``.
+Feel free to add new private/protected functions to the ``SetAssociativeCache``, but **DO NOT MODIFY THE PUBLIC INTERFACE**.
+
+You implementation of the set-associative cache should behave *exactly the same* as the direct-mapped cache when there is only one way.
+It should also be able to be any associative up to fully-associative (unless you explain why this is not possible as described below).
+
+You may choose to implement any replacement policy.
+**However, if your replacement policy requires any state, you must store this state in the tag array!**
+Note: The tag-array state is limited to 32 bits per line.
+If this constrains your associativity, be sure to include assertions in your code and explain this in the README file (see below).
+
+You **must** use the given TagArray and DataArray objects to store your tags and data.
+
+Implementing a Non-Blocking Cache
+---------------------------------
+
+Your non-blocking cache should inherit from your set-associative implementation as it must be both non-blocking and set-associative.
+Much of the code will be shared between the set-associative and the non-blocking cache.
+
+We discussed non-blocking caches in class.
+To implement a non-blocking cache, you must track outstanding requests from the cache.
+You may choose any software implementation of this you would like.
+However, you will be expected to explain how your software implementation would translate to hardware in interactive grading.
+
+Testing and Grading
+===================
+
+You should expect that we will test the following things:
+- Changing the size, associativity, and number of MSHRs of the cache.
+- Changing the line size of the memory.
+- Sending requests of any power-of-two size to the cache from the processor that are less than or equal to the line size.
+- Changing the address width.
+
+Note, we may use different implementations of ``TagArray``, ``SRAMArray``, ``Memory``, and ``Processor`` when testing.
+The only thing you can rely on is the interfaces in the header files.
+The underlying implementation may change.
+
+You should perform simulations with your code to answer the following questions.
+Include the answers to these questions in your README when submitting your assignment.
+**You should include specific data that back up your answers**.
+
+- How can you determine the hit ratio of the cache?
+- How does increasing the set-associativity from direct-mapped to 8-way set-associative affect the hit ratio?
+- How can you determine the performance of the system?
+- How does increasing the set-associativity from direct-mapped to 8-way set-associative affect the performance of the system?
+- How does increasing the number of MSHRs affect the performance of the system?
```

Comment:
What is a MSHR? You never defined it in this file. I assumed that this line should be "MSHR" instead of "MSRH" since all the other times you used it were the former.

Reply:
Yeah. Not sure why spellcheck didn't catch that...

MSHR: Miss status handling register.